src/router.ts

```diff
@@ -119,69 +118,81 @@ export function findRouteGivenExactInput(
   startToken: string,
   endToken: string,
   amountIn: number,
-  maxHops = 3,
+  maxHops: number,
 ): Route | null {
   const tokens = Object.keys(graph);
+  // distances[token][hop] is the maximum amount of token that can be received given hop number
   let distances: Distances = {};
+  // predecessors[token][hop] is the previous hop of the optimal path
   let predecessors: Predecessors = {};
-  let hops: Hops = {};
 
+  const defaultDistance = -Infinity;
   for (const token of tokens) {
-    distances[token] = -Infinity;
-    predecessors[token] = null;
-    hops[token] = 0;
+    distances[token] = {};
+    predecessors[token] = {};
   }
-  distances[startToken] = amountIn;
+  distances[startToken][0] = amountIn;
 
   for (let i = 0; i < maxHops; i++) {
     const newDistances = { ...distances };
```

Comment:
does `...` really deep clone the `Record<string, Record<number, number>>`? would be great to double-check.

to deep clone we can do `JSON.parse(JSON.stringify(distances))`

Reply:
It's a shallow copy - turns out we can skip the copy entirely